README.md
# RBHC
<p align="center">
<img src="https://github.com/intuit/RBHC/blob/master/Images/RBHC.png" width="400" height="400">
</p>

[![CircleCI](https://circleci.com/gh/intuit/RBHC/tree/master.svg?style=svg&circle-token=8dc91b504991b931f05d3b116040a4dfee4d0586)](https://circleci.com/gh/intuit/RBHC/tree/master)
[![codecov](https://codecov.io/gh/intuit/RBHC/branch/master/graph/badge.svg)](https://codecov.io/gh/intuit/RBHC)
[![PyPI version](https://badge.fury.io/py/RBHC.svg)](https://badge.fury.io/py/RBHC)
## Recursive Binary Hierarchical Clustering
This code is for accomplishing recursive binary hierarchical clustering of data <br>
K-Means algorithm is applied on the initial dataset and a binary partition is created after which using chi square score statistic, the feature (event) that was responsible for the partition is found out. The remaining clusters are further divided recursively using the above approach until the cluster size reaches 1 or the silhouette score reaches the threshold value <br>
- [Installation](#installation)
- [Usage](#usage)
- [Statistics](#statistics)
- [Visualisation](#visualisation)
- [Data File Structure](#data-file-structure)
- [Contribution and license](#contribution-and-license)
### Installation
Prerequisites: python3 <br>
```
pip install RBHC
```

### Usage
```
from RBHC import clustering
clustering(dataFilePath,thresholdValue)
```

  - dataFilePath = Path to data file  [Check data file structure](#data-file-structure)
  - thresholdValue = Silhouette value threshold (optional parameter and default in program is 0.65)

Return value from this function is a json with a tree structure that is generated with following important fields
  - name = Name of cluster node (string)
  - parent = Name of it's parent node (string)
  - size = Size of cluster (integer)
  - children = Tree structure of subtree (List)
  - clusterCreated = If clustering has been successful (Boolean)

To see a sample of this return value run clustering over sample dataset provided and print output or check visualisation/sampleData.json <br>

If you want to run this program in an interactive manner in a jupyter notebook run this command in root directory *jupyter notebook* and then it opens up in localhost

### Statistics
Once program runs then clustering statistics are stored in statistics/hierarchical/nameOfDataFile/ and for each sub cluster created stats are stored in a .json file and attributes are following

- ClusterId = Identifier of a sub cluster L=Level G=Number of cluster in that level counted left to right
- Size = Size of cluster
- Primary feature cluster created by = Name of feature which is responsible primarily for this cluster formation
- Features chi score = Shows chi score of all features in that cluster
- Stats on cluster by each feature = Stats of each feature in this cluster
- Ids = All instances that are part of cluster and names are derived from column[0] of data file

### Visualisation
Copy visualisation folder to directory where clustering is being used<br>
In visualisation folder nameOfDataFile.json will be created for clustering visualisation <br>
Run this in visualisation folder *python -m http.server 8888* and then in web browser open http://localhost:8888/
![](Clustering.gif)

### Data File Structure

```
IDS         | feature1    |                     | featureN
------------|-------------|---------------------|-----------------
ID1         |  value1     |                     |  valueN
            |             |                     |  
            |             |                     |
            |             |                     |  
```

All data files should be stored in data folder and check data folder for a sample .csv data file

### Contribution and license
- [Contributing guidelines](.github/CONTRIBUTING.md)
- [License](LICENSE)


CHANGELOG.md
# v1.0.0 (Saturday March 21 2020)

####

- README.md updated
- Unit tests added
- Clustering function returns a json structure of entire cluster tree
- CICD with circleci
- Code coverage on codecov
- Visualisation using d3.js
- Jupyter Support
- Integration with https://pypi.org/
- Clustering statistics added

#### Authors: 1

- Ashwith Atluri (Ashwith_Atluri@intuit.com)

# v1.0.1 (Saturday April 22 2020)

####

- README.md updated
- Codecov support added

#### Authors: 1

- Ashwith Atluri (Ashwith_Atluri@intuit.com)


Tests/tests.py
import unittest
import os
import sys
sys.path.append("../RBHC/")
from Clustering import *
clusterList={}
parentCluster=[[1,2,3,4,5,6,7,8,9],[9,9,9,9,9,9,9,9,9],[9,9,9,9,9,9,9,9,9],[9,9,9,9,9,9,9,9,9],[9,9,9,9,9,9,9,9,9]]
level=1
positions=[0,1,2,3,4]
parent='L0G0'
silhouetteThreshold=0.65
statsPath='Statisticstests/Hierarchical/unittest/'
user=['test','checking2','checking3','checking4','checking5']
attributes=['A','B','C','D','E','F','G','H','I']
globalCount={}
visualisationDirectory="Visualisationtests/"
modelNameFilePath="Visualisationtests/models.txt"
sampleJson={'name': 'L0G0', 'parent': 'null', 'desc': 'Cluster Size :6394', 'size': 6394, 'line_color': '#000', 'alert_color': '#000', 'children': [{'name': 'L1G0', 'desc': 'L1G0,Cluster Size :6335,Split by :Feature15,Mean : 2.3503', 'parent': 'L0G0', 'size': 6335, 'line_color': '#000', 'alert_color': '#000', 'children': [{'name': 'L2G0', 'desc': 'L2G0,Cluster Size :6163,Split by :Feature12,Mean : 2.0805', 'parent': 'L1G0', 'size': 6163, 'line_color': '#000', 'alert_color': '#000', 'children': [{'name': 'L3G0', 'desc': 'L3G0,Cluster Size :6156,Split by :Feature1,Mean : 0.4977', 'parent': 'L2G0', 'size': 6156, 'line_color': '#000', 'alert_color': '#000', 'children': [{'name': 'L4G0', 'desc': 'L4G0,Cluster Size :5977,Split by :Feature12,Mean : 0.9587', 'parent': 'L3G0', 'size': 5977, 'line_color': '#000', 'alert_color': '#000', 'children': [{'name': 'L5G0', 'desc': 'L5G0,Cluster Size :19,Split by :Feature1,Mean : 61.1053', 'parent': 'L4G0', 'size': 19, 'line_color': 'red', 'alert_color': 'red', 'children': [{'name': 'L6G0', 'desc': 'L6G0,Cluster Size :11,Split by :Feature1,Mean : 42.2727', 'parent': 'L5G0', 'size': 11, 'line_color': '#000', 'alert_color': '#000', 'children': []}, {'name': 'L6G1', 'desc': 'L6G1,Cluster Size :8,Split by :Feature1,Mean : 87.0', 'parent': 'L5G0', 'size': 8, 'line_color': '#000', 'alert_color': '#000', 'children': []}]}, {'name': 'L5G1', 'desc': 'L5G1,Cluster Size :5958,Split by :Feature1,Mean : 0.2969', 'parent': 'L4G0', 'size': 5958, 'line_color': '#000', 'alert_color': '#000', 'children': [{'name': 'L6G2', 'desc': 'L6G2,Cluster Size :5648,Split by :Feature12,Mean : 0.4232', 'parent': 'L5G1', 'size': 5648, 'line_color': '#000', 'alert_color': '#000', 'children': [{'name': 'L7G0', 'desc': 'L7G0,Cluster Size :204,Split by :Feature1,Mean : 2.3676', 'parent': 'L6G2', 'size': 204, 'line_color': 'red', 'alert_color': 'red', 'children': []}, {'name': 'L7G1', 'desc': 'L7G1,Cluster Size :5444,Split by :Feature1,Mean : 0.1964', 'parent': 'L6G2', 'size': 5444, 'line_color': '#000', 'alert_color': '#000', 'children': []}]}, {'name': 'L6G3', 'desc': 'L6G3,Cluster Size :310,Split by :Feature12,Mean : 10.7645', 'parent': 'L5G1', 'size': 310, 'line_color': 'red', 'alert_color': 'red', 'children': []}]}]}, {'name': 'L4G1', 'desc': 'L4G1,Cluster Size :179,Split by :Feature12,Mean : 39.6201', 'parent': 'L3G0', 'size': 179, 'line_color': 'red', 'alert_color': 'red', 'children': []}]}, {'name': 'L3G1', 'desc': 'L3G1,Cluster Size :7,Split by :Feature1,Mean : 221.8571', 'parent': 'L2G0', 'size': 7, 'line_color': 'red', 'alert_color': 'red', 'children': []}]}, {'name': 'L2G1', 'desc': 'L2G1,Cluster Size :172,Split by :Feature12,Mean : 124.6919', 'parent': 'L1G0', 'size': 172, 'line_color': 'red', 'alert_color': 'red', 'children': []}]}, {'name': 'L1G1', 'desc': 'L1G1,Cluster Size :59,Split by :Feature15,Mean : 420.2034', 'parent': 'L0G0', 'size': 59, 'line_color': 'red', 'alert_color': 'red', 'children': []}], 'clusterCreated': True}

class TestClustering(unittest.TestCase):
    def test_initializeClusterList(self):
        if(not os.path.exists(statsPath)):
            os.makedirs(statsPath)
        if(not os.path.exists(visualisationDirectory)):
            os.makedirs(visualisationDirectory)
        self.assertEqual(initializeClusterList(parentCluster,clusterList),True)
        self.assertEqual(hierarchicalClustering(parentCluster,level,positions,parent,clusterList["children"],silhouetteThreshold,statsPath,user,attributes,globalCount),True)
        self.assertEqual(writeCluster(visualisationDirectory+'unittest',clusterList,globalCount),True)
        self.assertEqual(createModelList(modelNameFilePath,'unittest'),True)
        self.assertEqual(clustering('../Data/sampleData.csv',0.65),sampleJson)
if __name__=='__main__':
    unittest.main()


Visualisation/js/clustering.js
var treeData = [];
// ************** Generate the dropdown menu adn i	 *****************
var filename,options;
var data=[] ;


var select = d3.select('.list')
       .append('select')
         .attr('class','select')
         .on('change',onchange)

d3.json("models.txt", function(error, list) {
      data=list
      data.unshift("Chose model")
      options = select
          .selectAll('option')
          .data(data).enter()
          .append('option')
          .text(function (d) { return d; });
});

function onchange() {
  selectValue = d3.select('select').property('value')
  filename=selectValue+".json"
  d3.select('svg').remove();
  d3.json(filename, function(error, treedata) {
    root = treedata[0];
    width=root['width']*300
    height=root['height']*240
    tree = d3.layout.tree()
       .size([width, height])
    svg = d3.select("body").append("svg")
       	.attr("width", width + margin.right + margin.left)
       	.attr("height", height + margin.top + margin.bottom)
         .append("g")
       	.attr("transform", "translate(" + margin.left + "," + margin.top + ")");
    root.x0 = width / 2
    root.y0 = 0;
    update(root);
    });
  d3.select('body')
  .append('p')

  };


// ************** Generate the tree diagram	 *****************
var margin = {top: 40, right: 120, bottom: 20, left: 260},
	width =500 - margin.right - margin.left,
	height = 200 - margin.top - margin.bottom;
var i = 0,
	duration = 750,
	root;
var nodeWidth = 10;
var nodeHeight = 300;
var horizontalSeparationBetweenNodes = 128;
var verticalSeparationBetweenNodes = 16;
var tree;
var svg;

var diagonal = d3.svg.diagonal()
	.projection(function(d) { return [d.x, d.y]; });


d3.select(self.frameElement).style("height", "500px");
function update(source) {
  // Compute the new tree layout.
  var nodes = tree.nodes(root).reverse(),
	  links = tree.links(nodes);
  // Normalize for fixed-depth.
  nodes.forEach(function(d) { d.y = d.depth * 200; });

  // Update the nodes…
  var node = svg.selectAll("g.node")
	  .data(nodes, function(d) { return d.id || (d.id = ++i); });

  // Enter any new nodes at the parent's previous position.
  var nodeEnter = node.enter().append("g")
	  .attr("class", "node")
	  .attr("transform", function(d) { return "translate(" + source.x0 + "," + source.y0 + ")"; })
	  .on("click", click);

  nodeEnter.append("circle")
	  .attr("r", 1e-6)
    .style("stroke", function(d) { return d.alert_color; })
	  .style("fill", function(d) { return d._children ? d.alert_color : "#fff"; });

  function wordwrap2(text) {
     var lines=text.split(",")
     return lines
  }
  nodeEnter.append("text")
	  .attr("text-anchor","start")
    .each(function (d) {
       if (d.desc!=undefined) {
          var lines = wordwrap2(d.desc)
          for (var i = 0; i < lines.length; i++) {
             d3.select(this).append("tspan")
             .attr("x",13)
           .attr("dy",20)
                  .text(lines[i])
           }
        }
      })
	  .style("fill-opacity", 1e-6);



  // Transition nodes to their new position.
  var nodeUpdate = node.transition()
	  .duration(duration)
	  .attr("transform", function(d) { return "translate(" + d.x + "," + d.y + ")"; });

  nodeUpdate.select("circle")
	  .attr("r", function(d) { return (18)})
    .style("stroke", function(d) { return d.alert_color; })
	  .style("fill", function(d) { return d._children ? d.alert_color : "#fff"; });

  nodeUpdate.select("text")
	  .style("fill-opacity", 1);

  // Transition exiting nodes to the parent's new position.
  var nodeExit = node.exit().transition()
	  .duration(duration)
	  .attr("transform", function(d) { return "translate(" + source.x + "," + source.y + ")"; })
	  .remove();

  nodeExit.select("circle")
	  .attr("r", 1e-6);

  nodeExit.select("text")
	  .style("fill-opacity", 1e-6);

  // Update the links…
  var link = svg.selectAll("path.link")
	  .data(links, function(d) { return d.target.id; });

  // Enter any new links at the parent's previous position.
  link.enter().insert("path", "g")
	  .attr("class", "link")
    .style("stroke", function(d) { return d.target.line_color; })
	  .attr("d", function(d) {
		var o = {x: source.x0, y: source.y0};
		return diagonal({source: o, target: o});
	  });

  // Transition links to their new position.
  link.transition()
	  .duration(duration)
	  .attr("d", diagonal);

  // Transition exiting nodes to the parent's new position.
  link.exit().transition()
	  .duration(duration)
	  .attr("d", function(d) {
		var o = {x: source.x, y: source.y};
		return diagonal({source: o, target: o});
	  })
	  .remove();

  // Stash the old positions for transition.
  nodes.forEach(function(d) {
	d.x0 = d.x;
	d.y0 = d.y;
  });
}

// Toggle children on click.
function click(d) {
  if (d.children) {
	d._children = d.children;
	d.children = null;
  } else {
	d.children = d._children;
	d._children = null;
  }
  update(d);
}


RBHC/Clustering.py
import pandas
import sklearn
import numpy
from sklearn import preprocessing
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
from sklearn.feature_selection import chi2
import json
import ast
import os
import glob
import operator
import argparse
import sys


def createModelList(modelNameFilePath, dataFile):
    ''' Appends model name to existing list of model name
    dataFile = name of data input file
    modelNameFilePath = path where models.txt is stored
    '''
    try:
        treeList = []
        write = True
        if (os.path.exists(modelNameFilePath)):
            fh = open(modelNameFilePath, "r")
            treeList = ast.literal_eval(fh.read())
            fh.close()
            if (str(dataFile)) not in treeList:
                treeList.append(dataFile)
            else:
                write = False
        else:
            treeList.append(dataFile)
        if (write):
            fh = open(modelNameFilePath, "w")
            fh.write(json.dumps(treeList, ensure_ascii='False'))
            fh.close()
    except Exception as e:
        print(e)
        return False
    return True


def writeCluster(fileName, clusterList, globalCount):
    ''' Writes cluster structure to json file
    fileName = name of data input file
    clusterList = hierarchical cluster created
    globalCount = dictionary with key as level and value as number of nodes at that level
    '''
    try:
        maximumKey = 0
        maximumValue = 0
        for key, value in globalCount.items():
            if (key > maximumKey):
                maximumKey = key
            if (value > maximumValue):
                maximumValue = value
        clusterList["height"] = maximumKey
        clusterList["width"] = maximumValue
        clusterListJson = json.dumps(clusterList, ensure_ascii='False', indent=4)
        fh = open(fileName + ".json", "w")
        fh.write("[" + clusterListJson + "]")
        fh.close()
    except Exception as e:
        print(e)
        return False
    return True


#recursive hierarchical clustering function
def hierarchicalClustering(parentCluster, level, positions, parent, clusterList,
                           silhouetteThreshold, statsPath, user, attributes, globalCount):
    ''' Recursive Binary Hierarchical Clustering Function
    parentCluster = cluster to be subclustered
    level = next level to be clustered
    positions = list of entities in parentCluster by index
    parent = name of parent cluster
    clusterList = hierarchical cluster structure to which subsequent clusters are appended to
    silhouetteThreshold = silhouette score threshold range in [-1 to 1]
    statsPath = path to statistics directory
    user = list of all user ids
    attributes = feature names
    globalCount = dictionary with key as level and value as number of nodes at that level
    '''
    print("At node " + parent)
    silhouetteScore = -1
    #tree node count dictionary at each level for labelling
    if (len(parentCluster) < 2):
        return False

    clusterCount = 2
    #clustering
    clusterer = KMeans(n_clusters=clusterCount, random_state=0)
    kmeans = clusterer.fit(parentCluster)
    clusterLabels = clusterer.fit_predict(parentCluster)
    # The silhouette_score gives the average value for all the samples.
    # This gives a perspective into the density and separation of the formed
    # clusters
    if (1 < len(set(clusterLabels)) < len(parentCluster)):
        silhouetteScore = silhouette_score(parentCluster, clusterLabels)

    #silhouette score threshold check
    if (silhouetteScore < silhouetteThreshold):
        return False

    indeX = chi2(parentCluster, clusterLabels)[0].tolist()
    #finding max chi score and its feature index
    chi_max = 0
    iterator = 0
    index_ = -1
    featureSignificance = {}
    while (iterator < len(indeX)):
        featureSignificance[attributes[iterator]] = indeX[iterator]
        if (indeX[iterator] > chi_max):
            chi_max = indeX[iterator]
            index_ = iterator
        iterator = iterator + 1
    featureSignificanceSorted = sorted(featureSignificance.items(),
                                       key=lambda x: x[1],
                                       reverse=True)
    try:
        globalCount[level]
    except:
        globalCount[level] = 0

    final_ = []
    childCluster = []
    {
        iterator: final_.append(numpy.where(kmeans.labels_ == iterator)[0])
        for iterator in range(kmeans.n_clusters)
    }
    clusterIterator = 0

    #for each cluster formed
    while clusterIterator < kmeans.n_clusters:
        iterator = 0
        childCluster = []
        new_positions = []
        ids = []
        while iterator < len(final_[clusterIterator]):
            childCluster.append(parentCluster[final_[clusterIterator][iterator]])
            new_positions.append(positions[final_[clusterIterator][iterator]])
            ids.append(user[positions[final_[clusterIterator][iterator]]])
            iterator = iterator + 1

        #cluster data interpretation
        clusterSummary = {}
        clusterName = 'L' + str(level) + 'G' + str(globalCount[level])
        dataIds = pandas.DataFrame(ids, columns=["IDS"])
        dataStats = pandas.DataFrame(childCluster, columns=attributes)
        stats = dataStats[attributes].head(n=len(childCluster)).describe()
        clusterSummary['ClusterId'] = clusterName
        clusterSummary['Size'] = len(childCluster)
        clusterSummary['Primary feature cluster created by'] = attributes[index_]
        clusterSummary['Features chi score'] = featureSignificanceSorted
        clusterSummary['Stats on cluster by each feature'] = stats.to_dict()
        clusterSummary['Ids'] = ids
        clusterSummaryJson = json.dumps(clusterSummary, ensure_ascii='False', indent=4)
        fh = open(statsPath + clusterName + 'clusterDescription' + ".json", "w")
        fh.write("[" + clusterSummaryJson + "]")
        fh.close()
        clusterIterator = clusterIterator + 1
        processingList = {}
        processingList["name"] = clusterName
        processingList["desc"] = clusterName + ",Cluster Size :" + str(
            len(childCluster)) + ',Split by :' + str(attributes[index_]) + ",Mean : " + str(
                round(dataStats[attributes[index_]].mean(), 4))
        processingList["parent"] = parent
        processingList['size'] = len(childCluster)

        if (float(len(childCluster) / len(parentCluster)) < 0.1):
            processingList['line_color'] = 'red'
            processingList['alert_color'] = 'red'

        else:
            processingList['line_color'] = '#000'
            processingList['alert_color'] = '#000'

        processingList["children"] = []
        globalCount[level] = globalCount[level] + 1
        hierarchicalClustering(childCluster, level + 1, new_positions, clusterName,
                               processingList["children"], silhouetteThreshold, statsPath, user,
                               attributes, globalCount)
        clusterList.append(processingList)
    return True


def initializeClusterList(transformedValues, clusterList):
    '''Initialises clusterList
    transformedValues = Input feature vector
    clusterList = hierarchical cluster structure to which subsequent clusters are appended to
    '''
    try:
        clusterList["name"] = 'L0G0'
        clusterList["parent"] = "null"
        clusterList["desc"] = "Cluster Size :" + str(len(transformedValues))
        clusterList["size"] = len(transformedValues)
        clusterList["line_color"] = '#000'
        clusterList["alert_color"] = '#000'
        clusterList["children"] = []
        clusterList["clusterCreated"] = False
    except Exception as e:
        print(e)
        return False
    return True


def clustering(dataFilePath, silhouetteThreshold=0.65):
    ''' Clusering function
    dataFilePath = Path of input csv file
    silhouetteThreshold = silhouette score threshold
    '''
    user = []
    transformedValues = []
    attributes = []
    clusterList = {}
    positions = []
    globalCount = {}
    if (not os.path.exists(dataFilePath)):
        print(dataFilePath + " Not found")
        sys.exit(-1)
    dataFile = os.path.splitext(os.path.basename(dataFilePath))[0]
    statsPath = "Statistics/Hierarchical/" + str(dataFile) + "/"
    visualisationDirectory = "Visualisation/"
    modelNameFilePath = visualisationDirectory + 'models.txt'
    vector = pandas.read_csv(dataFilePath)
    print(vector.head(n=len(vector)).describe())
    attributes = list(vector.columns[1:])
    transformedValues = vector.iloc[:, 1:].values.tolist()
    user = vector.iloc[:, 0].values.tolist()
    positions = list(range(0, len(transformedValues)))
    # Stats Path to be created
    if not os.path.exists(statsPath):
        os.makedirs(statsPath)
    else:
        files = glob.glob(statsPath + '/*')
        for f in files:
            os.remove(f)

    if (initializeClusterList(transformedValues, clusterList)):
        if (hierarchicalClustering(transformedValues, 1, positions, 'L0G0', clusterList["children"],
                                   silhouetteThreshold, statsPath, user, attributes, globalCount)):
            clusterList['clusterCreated'] = True
            if (os.path.isdir(visualisationDirectory)):
                if (writeCluster(visualisationDirectory + dataFile, clusterList, globalCount)):
                    createModelList(modelNameFilePath, dataFile)

    return (clusterList)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--datafilepath', help='Enter name of data file in data folder')
    parser.add_argument('--silhouettethreshold', help='Enter silhouetteThreshold')
    args = parser.parse_args()
    if (not args.datafilepath):
        print("Please Provide Data File")
        sys.exit(-1)
    dataFilePath = args.datafilepath
    if (not (args.silhouettethreshold)):
        silhouetteThreshold = 0.65
    else:
        silhouetteThreshold = args.silhouettethreshold
    clustering(dataFilePath, silhouetteThreshold)


if __name__ == "__main__":
    main()
